design/application.go

```diff
@@ -0,0 +1,177 @@
+		})
+
+		Result(ArrayOf(Application), "Return all registered Applications matching the query.")
+
+		Error("NotFound", func() {
+			Description("If the Application is not found, should return 404 Not Found.")
+		})
+
+		HTTP(func() {
+			GET("/applications")
+			Param("type", String, "List only Applications of given type", func() {
+				Example("predictor")
+			})
+			Param("workflow", String, "List only Applications generated by given workflow", func() {
+				Example("mlflow-sklearn-e2e")
+			})
+			Response(StatusOK)
+			Response("NotFound", StatusNotFound)
+		})
+
+		GRPC(func() {
+			// Responses use a "OK" gRPC code.
+			// The result is encoded in the response message (default).
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+		})
+
+	})
+
+	Method("register", func() {
+		Description("Register an Application with the FuseML application store.")
+
+		Payload(Application)
+
+		Error("BadRequest", func() {
+			Description("If the Application does not have the required fields, should return 400 Bad Request.")
+		})
+
+		Result(Application)
+
+		HTTP(func() {
+			POST("/applications")
+			Response(StatusCreated)
+			Response("BadRequest", StatusBadRequest)
+		})
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+	Method("get", func() {
+		Description("Retrieve an Application registered by FuseML.")
+
+		Payload(func() {
+			Field(1, "name", String, "Application name", func() {
+				Example("mlflow-seldon-predictor-01")
+			})
+			Required("name")
+		})
+
+		Error("BadRequest", func() {
+			Description("If name is not given, should return 400 Bad Request.")
+		})
+		Error("NotFound", func() {
+			Description("If there is no application with the given name, should return 404 Not Found.")
+		})
+
+		Result(Application)
+
+		HTTP(func() {
+			GET("/applications/{name}")
+			Response(StatusOK)
+			Response("BadRequest", StatusBadRequest)
+			Response("NotFound", StatusNotFound)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+			Response("NotFound", CodeNotFound)
+		})
+	})
+
+	Method("delete", func() {
+		Description("Delete an Application registered by FuseML application store.")
+
+		Payload(func() {
+			Field(1, "name", String, "Application name", func() {
+				Example("mlflow-seldon-predictor-01")
+			})
+			Field(2, "with_codeset", Boolean, "Delete related Codeset too", func() {
+				Default(false)
+			})
+			Field(3, "with_workflow", Boolean, "Delete related Workflow too", func() {
+				Default(false)
+			})
+			Required("name")
+		})
+
+		Error("BadRequest", func() {
+			Description("If the Application does not have the required fields, should return 400 Bad Request.")
+		})
+
+		HTTP(func() {
+			DELETE("/applications/{name}/")
+			Param("with_codeset", Boolean, "Delete related Codeset too", func() {
+				Default(false)
+			})
+			Param("with_workflow", Boolean, "Delete related Workflow too", func() {
+				Default(false)
+			})
```

Comment:
I'm still not sure what to think about these two parameters. I know you've been talking about adding them, and I was reluctant to say anything about them then, hoping I would make up my mind in the meantime.

On one hand, this makes applications very special among workflow outputs, because, in a sense, they will be perceived as the "owners" of codesets and workflows. I have a whole range of arguments against the idea that applications are special workflow outputs that should support this type of cascaded-delete operations:

1. some workflows can have no output applications at all (i.e. a workflow may only create other types of artifacts: models, datasets, even codesets, but no applications). Does that mean they are without owner ?
2. some workflows can have two or more applications as output. In this case, which one is the "owner" ?
3. some workflows can have more than codesets as input, or even no codesets as input at all (i.e. a workflow may have models or datasets, or even applications as input). What happens to those artifacts when you delete the application ? It doesn't make sense to model a `with_...` for every type of artifact that may be used as input
4. deleting codesets, in particular, doesn't make sense, because those are currently the objects most loosely coupled to workflows and, as an extension, to applications. The fact that we have a "codeset assign" operation should tell you that codesets are independent objects, able to exist on their own, and that we should prioritize modelling a "codeset unassign" operation over deleting codesets because they are indirectly related to some application
5. we need more than just a localized deletion operation, we need a unified strategy that applies to all artifacts of all types and sizes. In that respect, these parameters belong in another, more wide encompassing PR, perhaps one that implements those issues on retention policies and garbage collection
6. finally, we don't have enough user data to make the assumption that this is useful enough to warrant modelling these parameters so early on 

On the other hand, I do understand the need to support some cascaded-delete operation, but I think this should be done not in the server, but in the CLI/UI. This looks more like an interactive process to me, where a user deletes an application or a workflow and is then asked "do you also want to delete this list of related objects ?", where the objects can be workflows, or codesets, or something else. 

I recommend that you don't add these parameters at this time. Alternatively, if you really want to implement a cascaded-delete operation in the early stages of fuseml, you should do it for workflows, not for applications, because workflows at least are directly related to all these objects: codesets, applications and artifacts in general

Reply:
OK, I will remove it for now, that's not a problem.

Just one comment though: 

_Alternatively, if you really want to implement a cascaded-delete operation in the early stages of fuseml, you should do it for workflows, not for applications, because workflows at least are directly related to all these objects: codesets, applications and artifacts in general_

In current implementation, workflow does not know about the application; on the other hand application knows which workflow was used to create it. Thus application is the artefact on top of the food chain and others (codesets, workflows etc.) are just means to the end.

Reply:
> In current implementation, workflow does not know about the application; on the other hand application knows which workflow was used to create it. Thus application is the artefact on top of the food chain and others (codesets, workflows etc.) are just means to the end.

Yes, that's one way of looking at it: the output artifacts are the top of the food chain.

This will become clear when we revisit the way we implement workflow management, especially around how artifact inputs and outputs, not just applications and codesets, must be handled in general. In my view, the orchestrator must keep track of all input artifacts that it's providing to and all output artifacts (including applications) that it's extracting from each container representing a workflow run step, because that's how we implement end-to-end lineage for artifacts. Where exactly that data is stored, whether in the workflow run record, or spread out in the artifact records, is still a matter of debate.